Groom format, Port demonstration IO to h5, and fix Episodicity

shelhamer opened the pull request:
Switch from pickling to hdf5 for loading and saving of demonstrations.
This is a more portable format.

TODO
- [x] decide/fix episodicity (loss of life vs. game over, etc.)
- [x] discard partial episodes (as poor estimates of return)
- [x] record terminal trace instead of boundary indices
- [x] ~~switch from lists to arrays (might give a speed-up)~~ nope, lists b.c. I'm lazy
- [x] batch h5 write for each episode? no, less natural for the demonstrator.
- [x] figure out if h5 can go faster. select latest library version. (could switch to `lzf` if need be.)

shelhamer committed:
port demonstration IO to h5

h5 storage makes the data independent of the Demonstration class and
Python in general for portability.

- separate recording time step from ending episode/recording terminal
- drop frame bank, can compress during h5 write
- language: frame -> state, game over -> terminal

shelhamer committed:
prefer latest hdf5 (might be faster)

erictzeng commented on demonstration.py:
```diff
@@ -1,78 +1,46 @@
-        self.action_set = action_set
-        self.frames = FrameBank()
+    def __init__(self):
+        self.states = []
```
Do you actually mean `states` as opposed to `frames`? In my mind, a `state` is a collection of stacked frames, whereas a frame is a single emulator screenshot.

The naming is also a tad inconsistent with say, L6.

shelhamer replied:
I hear you, but I'd rather make this state b.c. it is the full observation at a timestep in the environment. That might suggest we should really call it `observation`, and let whatever MDP implementation define `state`, but I like the `(s, a, r)` naming.

L6 was a mistake.

erictzeng replied:
Sure, sounds good.

erictzeng commented on demonstration.py:
```diff
@@ -113,5 +81,11 @@ def reset_to_latest_snapshot(self, ale):
-            return pickle.load(f)
+        demo = Demonstration()
+        with h5py.File(path, 'r', libver='latest') as f:
+            demo.states = [s for s in np.array(f['S'])]
```
Talked with you a bit about this via Slack, but this still seems kind of strange to me...

erictzeng commented on demonstration.py:
```diff
@@ -1,78 +1,46 @@
+        return Timestep(self.states[index], self.actions[index],
+                        self.rewards[index], self.terminals[index])
 
     def save(self, path):
```
Is it worth also saving relevant metadata? The name of the ROM, the minimal action set, etc.?

I think saving the ROM name is at least a good idea. I vaguely recall having a reason for saving the minimal action set, but I think that was more of an ergonomic choice than an actual hard requirement, so I feel less strongly about that one.

shelhamer replied:
Agreed about rom name, will fix. Actions can be recovered by loading the rom. I expect the rom name will always be in the filename, but this makes it safe for the stupid filenames enforced by ALE.

erictzeng replied:
Yeah, the ergonomic concern was just that if you're training say, behavioral cloning, you don't actually need an instantiation of an ALE for anything beyond getting the minimal action set, so saving the action set makes it slightly easier.

This really isn't a big deal either way, though, so I'll defer to whatever you end up deciding.

shelhamer replied:
Ok, actions are included too as `action_set`.

erictzeng commented on record.py:
```diff
@@ -112,11 +112,11 @@ def record(ale, demo, output, num_frames, num_episodes):
-            demo.record_timestep(frame, action, reward, False)
+            demo.record_timestep(frame, action, reward)
             game_over = ale.game_over()
             if game_over:
```
Maybe inline the call to `ale.game_over()` if you get the chance...

shelhamer replied:
Leaving this for now, since I hope to replace it w/ a more complicated expression that captures lost lives too.

erictzeng commented on demonstration.py:
```diff
@@ -1,78 +1,46 @@
-        with open(path, 'wb') as f:
-            pickle.dump(self, f, pickle.HIGHEST_PROTOCOL)
+        with h5py.File(path, 'w', libver='latest') as f:
+            S = f.create_dataset('S', (len(self), ) + self.states[0].shape, dtype='uint8', compression='gzip', data=np.array(self.states))
```
If performance is bad, you could try switching to `lzf` instead of `gzip`, which should speed things up at the cost of worse compression ratios.

shelhamer committed:
record ROM name

shelhamer committed:
record action set

list ALE action indices for mapping to {0, ..., |A|-1} without invoking
the emulator

shelhamer committed:
h5: identify related fields

shelhamer committed:
handle snapshot interval in record

less coupling

shelhamer committed:
resume: fix ROM loading

the ROM path is required; the rom name from the demo does no good

shelhamer committed:
only record complete episodes

incomplete episodes are troubling b.c. they are potentially a poor
estimate of the return and because a demonstrator may not be able to
reasonably resume play during an episode.

shelhamer committed:
end episode on game over or loss of life

follow convention set by DeepMind

caution! the beam rider ROM has a bug that falsely reports lost lives.
DO NOT record demonstrations for beam rider for now.
